File: backend/app/schemas/device.py
from pydantic import BaseModel
from typing import Optional
from datetime import datetime, date

# Base schema for device attributes
class DeviceBase(BaseModel):
    name: str
    ip_address: str
    vendor: str
    username: str
    description: Optional[str] = None
    ha_peer_ip: Optional[str] = None
    use_ssh_for_last_hit_date: Optional[bool] = False
    collect_last_hit_date: Optional[bool] = True
    model: Optional[str] = None
    group: Optional[str] = None

    # 상세 정보
    serial_number: Optional[str] = None
    os_name: Optional[str] = None
    os_version: Optional[str] = None
    install_date: Optional[date] = None

    # 설치 위치
    location_region: Optional[str] = None
    location_building: Optional[str] = None
    location_floor: Optional[str] = None
    location_room: Optional[str] = None
    location_x: Optional[str] = None
    location_y: Optional[str] = None
    location_z: Optional[str] = None

    # 객체 수 임계치 (수기 입력, 사용량은 cached_* 컬럼과 비교)
    policy_threshold: Optional[int] = None
    network_object_threshold: Optional[int] = None
    service_threshold: Optional[int] = None

# Schema for creating a new device
class DeviceCreate(DeviceBase):
    password: str
    password_confirm: str

# Schema for updating an existing device (모든 필드 optional — PATCH 스타일 부분 업데이트 지원)
class DeviceUpdate(BaseModel):
    name: Optional[str] = None
    ip_address: Optional[str] = None
    vendor: Optional[str] = None
    username: Optional[str] = None
    description: Optional[str] = None
    ha_peer_ip: Optional[str] = None
    use_ssh_for_last_hit_date: Optional[bool] = None
    collect_last_hit_date: Optional[bool] = None
    model: Optional[str] = None
    group: Optional[str] = None
    password: Optional[str] = None
    password_confirm: Optional[str] = None

    serial_number: Optional[str] = None
    os_name: Optional[str] = None
    os_version: Optional[str] = None
    install_date: Optional[date] = None

    location_region: Optional[str] = None
    location_building: Optional[str] = None
    location_floor: Optional[str] = None
    location_room: Optional[str] = None
    location_x: Optional[str] = None
    location_y: Optional[str] = None
    location_z: Optional[str] = None

    policy_threshold: Optional[int] = None
    network_object_threshold: Optional[int] = None
    service_threshold: Optional[int] = None

# Schema for reading device data (from DB)
class Device(DeviceBase):
    id: int
    last_sync_at: Optional[datetime] = None
    last_sync_status: Optional[str] = None
    last_sync_step: Optional[str] = None

    # 대시보드 통계 캐시 (읽기 전용 — 동기화 완료 시 서버에서 갱신)
    cached_policies: Optional[int] = None
    cached_active_policies: Optional[int] = None
    cached_disabled_policies: Optional[int] = None
    cached_network_objects: Optional[int] = None
    cached_network_groups: Optional[int] = None
    cached_services: Optional[int] = None
    cached_service_groups: Optional[int] = None

    class Config:
        from_attributes = True

class DeviceSyncStatus(BaseModel):
    last_sync_at: Optional[datetime] = None
    last_sync_status: Optional[str] = None
    last_sync_step: Optional[str] = None

    class Config:
        from_attributes = True


class DeviceStats(BaseModel):
    """장비별 통계 정보"""
    id: int
    name: str
    vendor: str
    ip_address: str
    policies: int = 0
    active_policies: int = 0
    disabled_policies: int = 0
    network_objects: int = 0
    network_groups: int = 0
    services: int = 0
    service_groups: int = 0
    sync_status: Optional[str] = None
    sync_step: Optional[str] = None
    sync_time: Optional[datetime] = None
    policy_threshold: Optional[int] = None
    network_object_threshold: Optional[int] = None
    service_threshold: Optional[int] = None


class DashboardStatsResponse(BaseModel):
    """대시보드 통계 응답"""
    total_devices: int
    active_devices: int
    total_policies: int
    total_active_policies: int
    total_disabled_policies: int
    total_network_objects: int
    total_network_groups: int
    total_services: int
    total_service_groups: int
    device_stats: list[DeviceStats]

File: frontend/src/types/device.ts
export interface Device {
  id: number
  name: string
  ip_address: string
  vendor: string
  username: string
  description: string | null
  ha_peer_ip: string | null
  use_ssh_for_last_hit_date: boolean
  collect_last_hit_date: boolean
  model: string | null
  group: string | null
  last_sync_at: string | null
  last_sync_status: string | null
  last_sync_step: string | null
  serial_number: string | null
  os_name: string | null
  os_version: string | null
  install_date: string | null
  location_region: string | null
  location_building: string | null
  location_floor: string | null
  location_room: string | null
  location_x: string | null
  location_y: string | null
  location_z: string | null
  policy_threshold: number | null
  policy_threshold_manual: boolean
  network_object_threshold: number | null
  network_object_threshold_manual: boolean
  network_group_threshold: number | null
  network_group_threshold_manual: boolean
  service_threshold: number | null
  service_threshold_manual: boolean
  service_group_threshold: number | null
  service_group_threshold_manual: boolean
  cached_policies: number | null
  cached_active_policies: number | null
  cached_disabled_policies: number | null
  cached_network_objects: number | null
  cached_network_groups: number | null
  cached_services: number | null
  cached_service_groups: number | null
}

export interface DeviceCreate {
  name: string
  ip_address: string
  vendor: string
  username: string
  password: string
  password_confirm: string
  ha_peer_ip?: string
  description?: string
  use_ssh_for_last_hit_date?: boolean
  collect_last_hit_date?: boolean
  model?: string
  group?: string
  serial_number?: string
  os_name?: string
  os_version?: string
  install_date?: string
  location_region?: string
  location_building?: string
  location_floor?: string
  location_room?: string
  location_x?: string
  location_y?: string
  location_z?: string
  policy_threshold?: number
  policy_threshold_manual?: boolean
  network_object_threshold?: number
  network_object_threshold_manual?: boolean
  network_group_threshold?: number
  network_group_threshold_manual?: boolean
  service_threshold?: number
  service_threshold_manual?: boolean
  service_group_threshold?: number
  service_group_threshold_manual?: boolean
}

export interface DeviceUpdate {
  name?: string
  ip_address?: string
  vendor?: string
  username?: string
  password?: string
  password_confirm?: string
  ha_peer_ip?: string
  description?: string
  use_ssh_for_last_hit_date?: boolean
  collect_last_hit_date?: boolean
  model?: string
  group?: string
  serial_number?: string
  os_name?: string
  os_version?: string
  install_date?: string
  location_region?: string
  location_building?: string
  location_floor?: string
  location_room?: string
  location_x?: string
  location_y?: string
  location_z?: string
  policy_threshold?: number
  policy_threshold_manual?: boolean
  network_object_threshold?: number
  network_object_threshold_manual?: boolean
  network_group_threshold?: number
  network_group_threshold_manual?: boolean
  service_threshold?: number
  service_threshold_manual?: boolean
  service_group_threshold?: number
  service_group_threshold_manual?: boolean
}

export interface DeviceStats {
  id: number
  name: string
  vendor: string
  ip_address?: string
  sync_time: string | null
  sync_status: string | null
  sync_step: string | null
  policies: number
  active_policies: number
  disabled_policies: number
  network_objects: number
  network_groups: number
  services: number
  service_groups: number
  policy_threshold: number | null
  network_object_threshold: number | null
  network_group_threshold: number | null
  service_threshold: number | null
  service_group_threshold: number | null
}

export interface DashboardStats {
  total_devices: number
  active_devices: number
  total_policies: number
  total_active_policies: number
  total_disabled_policies: number
  total_network_objects: number
  total_network_groups: number
  total_services: number
  total_service_groups: number
  device_stats: DeviceStats[]
}
